README.md
# IDE-Haskell REPL ![](https://david-dm.org/atom-haskell/ide-haskell-repl.svg)

***This package is in beta.***

This package provides a way to interact with ghci (or, more precisely, `cabal repl` or `stack ghci`) from Atom.

![image](https://cloud.githubusercontent.com/assets/7275622/24383478/c3ae1ae4-1365-11e7-88a6-0c4877b23fc1.png)

## Documentation

Please refer to documentation site https://atom-haskell.github.io. In particular, see [here](https://atom-haskell.github.io/extra-packages/ide-haskell-repl/#using-on-windows) when running `ide-haskell-repl` on Windows.

## License

Copyright © 2015 Atom-Haskell.

Contributors (by number of commits):

<!-- BEGIN CONTRIBUTORS LIST -->
* Nikolay Yakimov
* yaeti
* Yakov Borevich
* soiamsoNG

<!-- END CONTRIBUTORS LIST -->

See [LICENSE.md][LICENSE] for details.

[LICENSE]: https://github.com/atom-haskell/ide-haskell-repl/blob/master/LICENSE.md


CHANGELOG.md
## 0.11.1

-   Spin event loop while waiting for repl output

    This should improve responsiveness in some cases.

-   Fix autocomplete box positioning

## 0.11.0

-   Bump atom-haskell-utils

    This should allow the package to work with cabalfiles up to 3.2, but it's a little experimental. Please open issues if stuff breaks with "Atom-Haskell core error".

## 0.10.2

-   Fix dangling ghci instances

## 0.10.1

-   Report ghci destruction on debug log, not as error
-   Avoid infinite recursion on bg repl destruction

## 0.10.0

-   Simplify error parsing somewhat
-   Fix duplicate error filtering
-   Colorize error severity in error div
-   Option to put errors in main repl output

## 0.9.7

-   Add run code button

## 0.9.5

-   Make error div selectable\/copyable
-   Update repl view when changing editor font size\/family

## 0.9.4

-   Enable setup-ghci-wrapper command on windows

## 0.9.3

-   Fix and clean-up wrapper set-up code

## 0.9.2

-   Add command to setup ghci-wrapper

## 0.9.1

-   Fix interrupt when using ghci-wrapper
-   Relax autoscroll condition

## 0.9.0

-   Reorder initialization to fix prompt bug
-   Bump dependencies & cabal-install 3.0 compatibility

## 0.8.7

-   Fix type collection on initial load

## 0.8.6

-   Optimize type collection a little bit

## 0.8.5

-   Don't relativize absolute paths in messages

## 0.8.4

-   Minor fixes, reload GHCI on no type (fixes no type tooltips until saved)

## 0.8.3

-   Update dependencies
-   Properly dispose of crashed background processes
-   Better spawn errors and crash handling

## 0.8.2

-   Fixed opening repl outside of a source file
-   Handle tabs in cabalfile more gracefully

    Note: tabs are forbidden in Cabalfiles, most of the ecosystem can't
    handle those. `stack` replaces tabs with spaces, but spews warnings.
    Other tools just silently fail. Please don't use tabs in cabalfiles.

-   Fix button state initialization

    Toggle buttons are now displayed correctly across Atom restarts

-   Bump dependencies

## 0.8.1

-   More careful asyncIterator definition
-   Avoid extraneous reload
-   Use atom-ts-spec-runner

## 0.8.0

-   Use reference-counting to track background REPL processes

    This avoids leaking REPL processes. This wasn't a big issue, since
    only one process per Cabal component is spawned, but no point in
    wasting resources.

-   More robust tracking of load/reload in REPL

    Manual loading/reloading via `:reload`, `:load` and `:edit` is now
    tracked.

-   Set editor to `atom --wait`

    This is mostly useful with newer Atom versions, where `--wait`
    actually waits before associated editor is closed

-   Better compilation error tracking and reporting

    Fixes \#65

-   Fixed show type ("absolutize" relative paths)

    GHCi 8.2.2 is somewhat inconsistent with how it gives file paths for
    `:all-types`. This problem is fixed.

-   Better reporting for startup errors
-   Support for cabal new-repl

## 0.7.16

-   Add busy indicator
-   Fix stdin input not showing up (\#62)
-   Fix spec runner
-   Pin typescript version
-   Bump dependencies

## 0.7.15

-   Fix error clear/show races, fix reload command

## 0.7.14

-   Do not allow input before REPL is initialized
-   Clear errors on reload

## 0.7.13

-   Don't send raw stdin if ghci is busy with autocompletion request

## 0.7.12

-   Display stdin input straight away; tweak style a bit

## 0.7.11

-   Prettier sources
-   Fix spec, Travis
-   Simplify focus retention, allow selecting output
-   Don't try to watch editor if repl view is destroyed
-   Remove repl editor registration in atom.textEditors

## 0.7.10

-   Atom 1.23 support (repl autocompletion broken)
-   Dispose of repl's TextEditor registration

## 0.7.9

-   Try hard to maintain editor focus
-   Update licensing information in README
-   Bump typings

## 0.7.8

-   Add tslib to dependencies

## 0.7.7

-   Bump Atom version
-   Try to focus text editor on repl creation; restore text editor focus
    on restart
-   New typings; update deps; bump es target
-   Added david-dm badge

## 0.7.6

-   Bump atom-haskell-utils (getComponentsFromFile bug)

## 0.7.5

-   Don't use --main-is, broken on new stack versions
-   Updates to lints, typings, minor fixes

## 0.7.4

-   Unset continuation prompt (GHC 8.2 compatibility)
-   Fix error rendering in absence of ide-haskell

## 0.7.3

-   Better reporting of startup errors

## 0.7.2

-   Fix EOL on Windows

## 0.7.1

-   Fallback on default repl in case of UPI-reladted errors
-   Enforce code style
-   Update typings

## 0.7.0

-   Rewrite in TypeScript
-   Use etch UI
-   Migrate to UPI 0.3
-   Updates for Atom 1.19
-   Bump min. Atom version to 1.19

## 0.6.0

-   Collect error messages
-   Add reload-repeat button
-   Refactor adding buttons
-   Button tooltips
-   Toggle autoReloadRepeat button
-   Serialize autoReloadRepeat
-   Add some repl-context commands
-   Some error handling for runCommand
-   Hide load/reload messages
-   Minimal indication that auto-reload-repeat is active
-   Common code, DRY
-   Factor autoReloadRepeat into replView
-   Add reload to editor context
-   Report pre-start errors

## 0.5.2

-   Use atom-highlight for highlighting

## 0.5.1

-   Merge remote-tracking branch 'origin/atom-1.13'
-   s/target/currentTarget/
-   Bump engine version
-   Atom 1.13 updates

## 0.5.0

-   GHCi-powered autocompletion in REPL input (by @yaeti)

## 0.4.9

-   Default output fonts

## 0.4.8

-   Fix selection commands

## 0.4.7

-   Merge pull request \#43 from soiamsoNG/patch-1
-   Quote the path (soiamsoNG)

## 0.4.6

-   Fix \#40 (error panel shrinking)

## 0.4.5

-   Fix some formatting issues on Windows
-   \[README\] reword "using on windows" a bit

## 0.4.4

-   Band-aid some Windows-specific issues

## 0.4.3

-   Revert "First attempt at fixing win32 interrupt"

## 0.4.2

-   First attempt at fixing win32 interrupt
-   Fix LICENSE date
-   Fix license

## 0.4.1


src/config.ts
export let config = {
  defaultRepl: {
    type: 'string',
    enum: ['stack', 'cabal-v1', 'ghci', 'cabal-v2'],
    default: 'ghci',
    order: 0,
  },
  stackPath: {
    type: 'string',
    default: 'stack',
    description: 'Path to stack executable',
    order: 10,
  },
  cabalPath: {
    type: 'string',
    default: 'cabal',
    description: 'Path to cabal executable',
    order: 20,
  },
  legacyCabalV1: {
    type: 'boolean',
    default: 'false',
    description:
      'Enable if using cabal-install < 2.4.0.0 and wish to use `cabal repl`',
    order: 21,
  },
  ghciPath: {
    type: 'string',
    default: 'ghci',
    description: 'Path to ghci executable',
    order: 30,
  },
  extraArgs: {
    type: 'array',
    default: [],
    description: 'Extra arguments passed to ghci. Comma-separated',
    items: {
      type: 'string',
    },
    order: 40,
  },
  autoReloadRepeat: {
    type: 'boolean',
    default: false,
    description: `Automatically reload and repeat last command on file save.
    This is only the default. You can toggle this per-editor using
    ide-haskell-repl:toggle-auto-reload-repeat command`,
    order: 50,
  },
  maxMessages: {
    type: 'number',
    default: 100,
    minimum: 0,
    description: `Maximum number of ghci messages shown. 0 means unlimited.`,
    order: 60,
  },
  errorsInOutput: {
    type: 'boolean',
    default: true,
    description: `Show interactive errors in output window.`,
    order: 65,
  },
  showTypes: {
    type: 'boolean',
    default: false,
    description: `Show type tooltips in ide-haskell if possible`,
    order: 70,
  },
  checkOnSave: {
    type: 'boolean',
    default: false,
    description: `Reload project in background when file is saved, will effectively
    check for errors`,
    order: 80,
  },
  ghciWrapperPath: {
    type: 'string',
    default: '',
    description: `This is intended to fix the 'interrupt closes ghci' problem
    on Windows -- see README for details. This option has no effect on
    other platforms`,
    order: 999,
  },
}

// generated by typed-config.js
declare module 'atom' {
  interface ConfigValues {
    'ide-haskell-repl.defaultRepl': 'stack' | 'cabal-v1' | 'ghci' | 'cabal-v2'
    'ide-haskell-repl.stackPath': string
    'ide-haskell-repl.cabalPath': string
    'ide-haskell-repl.legacyCabalV1': boolean
    'ide-haskell-repl.ghciPath': string
    'ide-haskell-repl.extraArgs': string[]
    'ide-haskell-repl.autoReloadRepeat': boolean
    'ide-haskell-repl.maxMessages': number
    'ide-haskell-repl.errorsInOutput': boolean
    'ide-haskell-repl.showTypes': boolean
    'ide-haskell-repl.checkOnSave': boolean
    'ide-haskell-repl.ghciWrapperPath': string
    'ide-haskell-repl': {
      defaultRepl: 'stack' | 'cabal-v1' | 'ghci' | 'cabal-v2'
      stackPath: string
      cabalPath: string
      legacyCabalV1: boolean
      ghciPath: string
      extraArgs: string[]
      autoReloadRepeat: boolean
      maxMessages: number
      errorsInOutput: boolean
      showTypes: boolean
      checkOnSave: boolean
      ghciWrapperPath: string
    }
  }
}
